### PR TITLE
Reshuffle cabal packages in ouroboros-network

### PR DESCRIPTION
**Pull request by coot**
Introduces:
* `typed-protocols-cbor` package which contains `PingPong` and `ReqResp`, also include tests of these codecs; This package is required for testing components of `mux` and `ouroboros-network`
* `mux` package with `Mux` modules, related `Channel` functions and bearers (`Network.Mux.Bearer.{Queues,Pipe,Socket}`; we could make it independent of `typed-protocols` by introducing its own  `Channel` type - this is left for later. 

I left the `Ouroboros.Network.Socket.Server` in `ouroboros-network`, since it's not related to `mux`.  This makes `Network.Mux.Bearer.{Queues,Pipe}` slightly different from `Network.Mux.Bearer.Socket`.  The former provide `runMuxBeaerWith{Queues,Pipe}`, the other does not.  The reason for that is that complexity of using a mux over listening sockects is much greater than simple `bracket` function around creating mux bearer as it is the case for `Queues` and `Pipe`s.

This PR addresses #677

### CODE REVIEW

File: mux/test/Test/Mux.hs [hunk 1]
```diff
@@ -0,0 +1,659 @@
+{-# LANGUAGE BangPatterns               #-}
```
Comment (karknu) [inline mux/test/Test/Mux.hs]: This is a rename of `ouroboros-network/test/Test/Mux.hs` so you should try and use `git mv` to preserve version history.

File: ouroboros-network/test/Test/Pipe.hs [hunk 2]
```diff
@@ -24,17 +24,18 @@ import           Test.Tasty.QuickCheck (testProperty)
...
 import qualified Ouroboros.Network.ChainProducerState as CPS
-import qualified Ouroboros.Network.Mux as Mx
-import qualified Ouroboros.Network.Mux.Interface as Mx
-import           Ouroboros.Network.Pipe
 import           Ouroboros.Network.Protocol.ChainSync.Client as ChainSync
 import           Ouroboros.Network.Protocol.ChainSync.Codec as ChainSync
 import           Ouroboros.Network.Protocol.ChainSync.Examples as ChainSync
 import           Ouroboros.Network.Protocol.ChainSync.Server as ChainSync
-import qualified Test.Mux as Mxt
+-- import qualified Test.Mux as Mxt
```
Comment (karknu) [inline ouroboros-network/test/Test/Pipe.hs]: Remove?

File: mux/test/Test/Mux.hs (hunk 1 above)
Comment (coot) [reply mux/test/Test/Mux.hs]: It's only part of it, the `demo` stayed in `ouroboros-network`

File: ouroboros-network/test/Test/Pipe.hs (hunk 2 above)
Comment (coot) [reply ouroboros-network/test/Test/Pipe.hs]: Yes, and all this should go to `mux` package too.

Comment (coot) [reply ouroboros-network/test/Test/Pipe.hs]: It cannot it is using `ChainSync` protocol.

**Comment by coot** [comment]
Thanks @karknu, Server.Socket` is independent of the mux so we can decide later where it should land.

**Pull request merged**